### PR TITLE
Ignore AUTO_INCREMENT in CREATE TABLE to simplify migration from MySQL

### PR DESCRIPTION
**Pull request by devcrafter**
Issue #35893

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Option `compatibility_ignore_auto_increment_in_create_table` allows ignoring `AUTO_INCREMENT` keyword in a column declaration to simplify migration from MySQL

### CODE REVIEW

File: tests/queries/0_stateless/02293_compatibility_ignore_auto_increment_in_create_table.sql [hunk 1]
```diff
@@ -0,0 +1,36 @@
+select 'disable AUTO_INCREMENT compatibility mode';
+set compatibility_ignore_auto_increment_in_create_table=false;
+
+select 'create table with AUTO_INCREMENT, compatibility disabled';
+DROP TABLE IF EXISTS ignore_auto_increment SYNC;
+CREATE TABLE ignore_auto_increment (
+    id int AUTO_INCREMENT
+) ENGINE=MergeTree() ORDER BY tuple(); -- {serverError SYNTAX_ERROR}
```
Comment (alesapin) [inline tests/queries/0_stateless/02293_compatibility_ignore_auto_increment_in_create_table.sql:8]: Maybe add case with `AUTO_INCREMENT` and/or `DEFAULT` and/or `EPHEMERAL` for a single column.

File: src/Interpreters/InterpreterCreateQuery.cpp [hunk 2]
```diff
@@ -570,7 +580,7 @@ ColumnsDescription InterpreterCreateQuery::getColumnsDescription(
         else if (col_decl.type)
             column.type = name_type_it->type;
         else
-            throw Exception();
+            throw Exception{"Neither default experssion nor type is provided for a column", ErrorCodes::LOGICAL_ERROR};
```
[the remark this answers is not in the document]
Comment (devcrafter) [reply src/Interpreters/InterpreterCreateQuery.cpp:583]: IMO, the execution shouldn't end up here. AFAIU, according to our code, a column should have either type or default expression. I end up here during development by playing with conditions, and it was weird to see an error with code 0

File: src/Parsers/ParserCreateQuery.h [hunk 3]
```diff
@@ -196,13 +198,25 @@ bool IParserColumnDeclaration<NameParser>::parseImpl(Pos & pos, ASTPtr & node, E
...
         if (!literal_parser.parse(pos, default_expression, expected) && type)
             default_expression = std::make_shared<ASTLiteral>(Field());
 
         if (!default_expression && !type)
             return false;
     }
+    else if (s_auto_increment.ignore(pos, expected))
+    {
+        default_specifier = s_auto_increment.getName();
+        /// if type is not provided for a column with AUTO_INCREMENT then using INT by default
```
[the remark this answers is not in the document]
Comment (devcrafter) [reply src/Parsers/ParserCreateQuery.h:211]: In MySQL, you have to specify type. In CH, it's not necessary to provide type if default expression is provided. So, it's done similarly.

Also, I see that there is `AUTO_INCREMENT` syntax with initial value, which I didn't handle here. I'll add it

[the remark this answers is not in the document]
Comment (devcrafter) [reply src/Parsers/ParserCreateQuery.h:211]: I misread it. There is no initial value for `AUTO_INCREMENT`

File: src/Interpreters/InterpreterCreateQuery.cpp [hunk 4]
```diff
@@ -552,6 +553,16 @@ ColumnsDescription InterpreterCreateQuery::getColumnsDescription(
 
         column.name = col_decl.name;
 
+        /// ignore or not other database extensions depending on compatibility settings
+        if (col_decl.default_specifier == "AUTO_INCREMENT"
+            && !context_->getSettingsRef().compatibility_ignore_auto_increment_in_create_table)
+        {
+            throw Exception{
+                "AUTO_INCREMENT is not supported. To ignore the keyword in column declaration set "
```
Comment (devcrafter) [inline src/Interpreters/InterpreterCreateQuery.cpp:561]: " ... declaration, set ..."

**Comment by alesapin** [comment]
Style check Ok.

**Pull request merged**